druntime/src/core/memory.d

```diff
@@ -278,7 +278,12 @@ extern(C):
      * process footprint.  Collections may continue to occur in instances
      * where the implementation deems necessary for correct program behavior,
      * such as during an out of memory condition.  This function is reentrant,
-     * but enable must be called once for each call to disable.
+     * but `GC.enable()` must be called once for each call to `GC.disable()`.
+     * Unlike the
+     * $(LINK2 https://dlang.org/spec/function.html#nogc-functions, `@nogc` attribute),
+     * `GC.disable()` halts
+     * collections across all threads, yet still allows GC allocations.
+     * Disabling collections eliminates GC pauses.
```

Comment:
I think maybe it would be good to start a discussion here.

As far as I understand it, calling `GC.disable()` will disable collections except in the case that the system is low on memory and the GC determines it could run a collection cycle to free up memory. That's somewhat covered by the doc `"Collections may continue to occur in instances where the implementation deems necessary for correct program behavior, such as during an out of memory condition."`. But I don't think people are paying too much attention here and assume `GC.disable()` is all they need.

I base this on the following code:
- [isLowOnMem()](https://github.com/dlang/dmd/blob/2a7450636892eb3696c3e1656fc5b15dcf537354/druntime/src/core/internal/gc/os.d#L213-L252). Note that the Posix platform implementation here needs to be fixed. It uses hardcoded values of 3GB or 4GB. I think these assumptions are dated for nowadays hardware (though that depends entirely on the target). We should probably backlog this -- and perhaps make it customizable via ENV vars.
- The [GC alloc code](https://github.com/dlang/dmd/blob/2a7450636892eb3696c3e1656fc5b15dcf537354/druntime/src/core/internal/gc/impl/conservative/gc.d#L1949-L1966) which will ignore the `disabled` state if the system is low on memory. In other words if `isLowOnMem()` returns true.

So in fact the GC will collect as a last resort.

I wonder though.. If the user's intention was to /never/ pause because it would be catastrophic in the domain of the system they designed then maybe the GC should throw an error rather than ignore the user's preference? At the very least it should be customizable -- perhaps with an ENV variable.

Reply:
I suggest filing this as a bugzilla and bringing it up in the n.g.